### PR TITLE
core, trie: rework trie database

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR reworks the trie database a bit to simplify things and make it be usable by verkle tree.

Trie database uses reference counter mechanism to prune stale nodes in memory. It relies on the references
between parent node and child nodes to determine if the node is already outdated.

However, the trie database is very MPT specific. For example, it maintains the MPT nodes directly and iterate
children of nodes with MPT rules. It will bring some performance speed up by avoiding node encoding/decoding.
But it's not abstract enough, and very unfriendly for new trie types, e.g. verkle tree.

In this PR, trie database will only maintain encoded node blobs with additional metadata(e.g the children hashes
node). In this way, trie database could be abstract enough to be used by different trie implementations. And also
the huge benefit is it unties the trie and trie database, we can move the trie database to a separate package
in the future.

The downsize is trie database needs to maintain the children hashes of nodes additionally, it will requires more
memory space. We need to twist a reasonable memory allowance for dirty node set.

### CODE REVIEW

File: trie/nodeset.go [hunk 1]
```diff
@@ -27,41 +26,29 @@ import (
 // memoryNode is all the information we know about a single cached trie node
 // in the memory.
 type memoryNode struct {
-	hash common.Hash // Node hash, computed by hashing rlp value, empty for deleted nodes
-	size uint16      // Byte size of the useful cached data, 0 for deleted nodes
-	node node        // Cached collapsed trie node, or raw rlp data, nil for deleted nodes
+	hash     common.Hash   // Node hash, computed by hashing rlp value, empty for deleted nodes
+	node     []byte        // Cached RLP-encoded trie node, nil for deleted nodes
+	children []common.Hash // The list of children hashes
```
[the remark this answers is not in the document]
Comment (rjl493456442) [reply trie/nodeset.go:31]: Sounds cool! 

Indeed as you said children list will be very large and would be way more
 better for taking byte offset approach. I will have a try.

[the remark this answers is not in the document]
Comment (fjl) [reply trie/nodeset.go:31]: One problem will be getting the actual offsets. I think this will require a rewrite/duplication of the node decoder functions.

[the remark this answers is not in the document]
Comment (fjl) [reply trie/nodeset.go:31]: Also, I do not advise implementing the generics-based thing right away, since that will further increase code complexity.

**Comment by rjl493456442** [comment]
截屏2023-03-13 下午4 15 43
截屏2023-03-13 下午4 17 03

After running 7 days, this PR is slightly slower than master branch, which is approximately 1.7% slower
because of additional RLP encoding/decoding overhead.

File: trie/database.go [hunk 2]
```diff
@@ -90,184 +96,37 @@ type Database struct {
...
-	size uint16 // Byte size of the useful cached data
-
-	parents  uint32                 // Number of live nodes referencing this one
-	children map[common.Hash]uint16 // External children referenced by this node
-
-	flushPrev common.Hash // Previous node in the flush-list
-	flushNext common.Hash // Next node in the flush-list
+	node      []byte                   // Encoded node blob
+	parents   uint32                   // Number of live nodes referencing this one
+	external  map[common.Hash]struct{} // The set of external children
```
Comment (karalabe) [inline trie/database.go:104]: Why the change from int to marker? Won't this lose a "count" of the reference?

File: trie/database.go [hunk 3]
```diff
@@ -293,34 +152,37 @@ func NewDatabaseWithConfig(diskdb ethdb.Database, config *Config) *Database {
 	if config != nil && config.Preimages {
 		preimage = newPreimageStore(diskdb)
 	}
-	db := &Database{
-		diskdb: diskdb,
-		cleans: cleans,
-		dirties: map[common.Hash]*cachedNode{{}: {
-			children: make(map[common.Hash]uint16),
-		}},
+	var resolver ChildResolver
```
Comment (karalabe) [inline trie/database.go:155]: Any particular reason for ChildResolver and MPTResolver to be public?

File: trie/database.go (hunk 2 above)
Comment (karalabe) [reply trie/database.go:104]: I guess one example could be the meta-parent (`common.Hash`) referencing the same root hash multiple times (Clique networks, post-merge empty-block state). In that case, we'll not add the second reference this way and the dereference of the first might actually delete data still referenced later?

Comment (karalabe) [reply trie/database.go:104]: I guess one example could be the meta-parent (`common.Hash`) referencing the same root hash multiple times (Clique networks, post-merge empty-block state). In that case, we'll not add the second reference this way and the dereference of the first might actually delete data still referenced later?

Comment (rjl493456442) [reply trie/database.go:104]: I don't maintain the common.Hash -> root references anymore, seems I found it's unnecessary.
The parents counter of root node is already necessary.

Comment (rjl493456442) [reply trie/database.go:104]: We can even replace it with a common.Hash pointer, because in our case, a node can have at most 1 external child(storage root).

File: trie/database.go (hunk 3 above)
Comment (rjl493456442) [reply trie/database.go:155]: Because I plan to move the trie.Database to a separate package(maybe `trie/triedb`) and put all pbss stuff there(maybe `trie/triedb/pbss`) as well. So I make this resolve publicly.

**Comment by karalabe** [comment]
Pls rebase

File: trie/committer.go [hunk 4]
```diff
@@ -142,12 +142,10 @@ func (c *committer) store(path []byte, n node) node {
 	// We have the hash already, estimate the RLP encoding-size of the node.
```
Comment (MariusVanDerWijden) [inline trie/committer.go:142]: This comment is stale now I think, we're not estimating the size anymore.
Also we could create a pool for rlp.NewEncodeBuffer within `nodeToBytes` to reduce allocs

File: trie/committer.go [hunk 5]
```diff
@@ -139,15 +139,11 @@ func (c *committer) store(path []byte, n node) node {
...
-	// We have the hash already, estimate the RLP encoding-size of the node.
-	// The size is used for mem tracking, does not need to be exact
 	var (
-		size  = estimateSize(n)
 		nhash = common.BytesToHash(hash)
 		mnode = &memoryNode{
 			hash: nhash,
-			node: simplifyNode(n),
-			size: uint16(size),
+			node: nodeToBytes(n),
```
[the remark this answers is not in the document]
Comment (rjl493456442) [reply trie/committer.go:146]: I think about it. After the change in this PR, there are some extra costs for verkle, mostly the node serialization and deserialization:

#### Scenario description

There are a few hundreds accounts and a few thousands touched in the mainnet block. Let's use 500 accounts and 3.5K slots here for performance estimation.

In verkle, the node width is 256, the average tree depth is 4. If we need to touch 4K leaves in total, it means 4K leaf nodes and 585(1+8+8^2+8^3) internal nodes will be accessed.

#### Deserialization
It happens when the node is accessed. The major cost is the conversion from node hash to node commitment(find the elliptic curve point by its X coordinator). I wrote a benchmark and it turns out it's indeed extremely expensive.

```go
// BenchmarkHashToCommitment
// BenchmarkHashToCommitment-8   	   45598	     25769 ns/op	      32 B/op	       1 allocs/o
func BenchmarkHashToCommitment(b *testing.B) {
	p := &Generator
	p.Add(p, p)
	hash := p.Bytes()

	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		var elem Element
		elem.SetBytesTrusted(hash[:])
	}
}
```

If we need to do the conversion for 4.58K nodes, the overall cost will be 112ms. Fortunately we have trie prefetcher which will try to load and decode nodes during evm execution. Not sure how much can be covered by it though.

#### Serialization

The same, the major cost of serialization comes from the conversion of node commitment to node hash. And it's possible to speed it up by running the conversion in a batch.

So the total cost of this conversion is 4.5ms and might be possible to reduce it a lot.

```go
// BenchmarkCommitmentToHash
// BenchmarkCommitmentToHash-8   	  740104	      1625 ns/op	       0 B/op	       0 allocs/op
func BenchmarkCommitmentToHash(b *testing.B) {
	p := &Generator
	p.Add(p, p)

	b.ResetTimer()
	b.ReportAllocs()

	for i := 0; i < b.N; i++ {
		p.Bytes()
	}
}

// nodes = 16, 384.625 ns in average
// BenchmarkElementsToBytes
// BenchmarkElementsToBytes-8   	  169038	      6154 ns/op	    1552 B/op	       4 allocs/op
func BenchmarkElementsToBytes16(b *testing.B) {
	benchmarkElementsToBytes(b, 16)
}

// nodes = 256, 250.41015625 ns in average
// BenchmarkElementsToBytes256
// BenchmarkElementsToBytes256-8   	   17974	     64105 ns/op	   24832 B/op	       4 allocs/op
func BenchmarkElementsToBytes256(b *testing.B) {
	benchmarkElementsToBytes(b, 256)
}
```

#### Summary

The cost of node serialization and deserialization is not trivial. But if we keep the node or a pointer of node in memory, it will require some additional abstraction and might be MPT or Verkle specific. I hope we can optimize a bit to reduce the cost.

[the remark this answers is not in the document]
Comment (rjl493456442) [reply trie/committer.go:146]: But as you said, if we change the serialization rules a bit, by retaining the commitment(elliptic curve point) itself instead of the hash(x coordination of point), then we can get rid of this part of cost. But the cost is node size is larger.

I think we really need to twist the trie storage scheme, since currently the verkle node can have at most 256 children and the size of child is 32B. The database amplification is already very large. We need to somehow split the children set to reduce the database amplification.

**Review by karalabe — APPROVED** [review]
SGTM, but really only time running will tell if everything's truly correct 🙃
